move spark integration page to focus on assets (#14537)
dagster-io/dagster: docs/content/integrations/spark.mdx

---
title: Dagster with Spark
description: Dagster can perform computations using Spark.
---

# Using Dagster with Spark

Dagster assets and ops can perform computations using Spark.

Running computations on Spark presents unique challenges, because, unlike other computations, Spark jobs typically execute on infrastructure that's specialized for Spark - i.e. that can network sets of workers into clusters that Spark can run computations against. Spark applications are typically not containerized or executed on Kubernetes. Running Spark code often requires submitting code to a Databricks or EMR cluster.

There are two approaches to writing Dagster assets and ops that invoke Spark computations:

## Asset body submits Spark job

With this approach, the code inside the asset definition submits a Spark job to an external system like Databricks or EMR, usually pointing to a jar or zip of Python files that contain the actual Spark data transformations and actions.

> If you want to run a Spark job against YARN or a Spark Standalone cluster, the code inside your asset can issue a shell command invokes `spark-submit`.

Reviewer: ```suggestion
If you want to run a Spark job against YARN or a Spark Standalone cluster, the code inside your asset can issue a shell command that invokes `spark-submit`.
```

Author: Thanks for catching - fixed this.

This is the easiest approach for migrating existing Spark jobs, and it's the only approach that works for Spark jobs written in Java or Scala. The downside is that it loses out on some of the benefits of Dagster - the implementation of each asset is bound to the execution environment, so you can't run your Spark transformations without relying on external infrastructure. Writing unit tests is cumbersome.

> ## Asset accepts and produces DataFrames or RDDs

Reviewer: What's an RDD? Can you define it before using it as an acronym?

> With this approach, the code inside the asset definition consists of pure logical data transformations on Spark DataFrames or RDDs. The `asset`-decorated function accepts `DataFrames` as parameters and returns `DataFrames` when it completes. An [IO manager](/concepts/io-management/io-managers) handles writing and reading the DataFrames to and from persistent storage. The [Running PySpark code in assets](#running-pyspark-code-in-assets) example below shows what this looks like.

Reviewer: > The `asset`-decorated function

Why use code formatting for asset over leaving it plain or working in the PyObject component for `@asset`?

Author: Good point. I fixed this to use PyObject.

> With this approach, the code inside the asset definition consists of pure logical data transformations on Spark DataFrames or RDDs. The `asset`-decorated function accepts `DataFrames` as parameters and returns `DataFrames` when it completes. An [IO manager](/concepts/io-management/io-managers) handles writing and reading the DataFrames to and from persistent storage. The [Running PySpark code in assets](#running-pyspark-code-in-assets) example below shows what this looks like.

Reviewer: > An [IO manager](/concepts/io-management/io-managers)

This should be `I/O manager`

> If you want your Spark driver to run inside a Spark cluster, you use a "step launcher" resource that informs Dagster how to launch the step. The step launcher resource is responsible for invoking `spark-submit` or submitting the job to Databricks or EMR. [Submitting PySpark jobs on EMR](#submitting-pyspark-jobs-on-emr) shows what this looks like for EMR.

Reviewer: ```suggestion
If you want your Spark driver to run inside a Spark cluster, you can use a "step launcher" resource that informs Dagster how to launch the step. The step launcher resource is responsible for invoking `spark-submit` or submitting the job to Databricks or EMR. [Submitting PySpark jobs on EMR](#submitting-pyspark-jobs-on-emr) shows what this looks like for EMR.
```

Also, why is step launcher in quotes?

> The advantage of this approach is a very clean local testing story. You can run an entire pipeline of Spark assets in a single process. You can use [IO managers](/concepts/io-management/io-managers) to abstract away IO - storing outputs on the local filesystem during local development and in the cloud in production.

Reviewer: ```suggestion
The advantage of this approach is a clean local testing story. You can run an entire pipeline of Spark assets in a single process. You can use [I/O managers](/concepts/io-management/io-managers) to abstract away I/O - storing outputs on the local filesystem during local development and in the cloud in production.
```

Suggesting we remove a qualifier (`very`) and format IO as I/O

The downside is that this approach only works with PySpark, and setting up a step launcher can be difficult. We currently provide an <PyObject module='dagster_aws.emr' object='emr_pyspark_step_launcher' /> and a <PyObject module='dagster_databricks' object='databricks_pyspark_step_launcher' />, but if you need to submit your Spark job to a different kind of cluster, writing your own can be time consuming ([here are some tips](https://github.com/dagster-io/dagster/discussions/3201)). You also need to install the Dagster library itself on the cluster.

### Running PySpark code in assets

<CodeReferenceLink filePath="examples/with_pyspark" />

Passing PySpark DataFrames between asset functions requires a little bit of extra care, compared to other data types, for a couple reasons:

- Spark has a lazy execution model, which means that PySpark won't process any data until an action like `write` or `collect` is called on a DataFrame.

> - PySpark DataFrames cannot be pickled, which means that [IO Managers](/concepts/io-management/io-managers) like the default filesystem IO manager won't work for them.

Reviewer: ```suggestion
- PySpark DataFrames cannot be pickled, which means that [I/O managers](/concepts/io-management/io-managers) like the default filesystem I/O manager won't work for them.
```

`IO` > `I/O`

In this example, we've defined an <PyObject module='dagster' object='IOManager' /> that knows how to store and retrieve PySpark DataFrames that are produced and consumed by ops.

> This example assumes that all the outputs within the Dagster pipeline will be PySpark DataFrames and stored in the same way. To learn how to use different IO managers for different outputs within the same Dagster pipeline, take a look at the [IO Manager](/concepts/io-management/io-managers) concept page.

Reviewer: ```suggestion
This example assumes that all the outputs within the Dagster pipeline will be PySpark DataFrames and stored in the same way. To learn how to use different IO managers for different outputs within the same Dagster pipeline, take a look at the [I/O manager](/concepts/io-management/io-managers) concept page.
```

I know this wasn't part of your changes, but can we please add it in?

> This example writes out `DataFrames` to the local file system, but can be tweaked to write to cloud object stores like S3 by changing to the `write` and `read` invocations.

Reviewer: ```suggestion
This example writes out `DataFrames` to the local file system, but can be tweaked to write to cloud object stores like Amazon S3 by changing to the `write` and `read` invocations.
```

The first time a branded product is mentioned, the name of the company/creator/provider/etc should be include. Like `Google BigQuery`, thereafter just `BigQuery`.
